llvm/lib/Target/AMDGPU/SIInstrInfo.cpp

```diff
@@ -151,6 +151,9 @@ static bool resultDependsOnExec(const MachineInstr &MI) {
   // Ignore comparisons which are only used masked with exec.
   // This allows some hoisting/sinking of VALU comparisons.
   if (MI.isCompare()) {
+    if (SIInstrInfo::isVOPC(MI))
```

Comment:
Probably should directly check the first operand is the only explicit def rather than assume it's implied by the encoding

Reply:
It is really not clear to me why this is correct. The function comment says "Returns true if the scalar result of a VALU instruction depends on exec" - but what does that mean for an instruction with no explicit def?

Reply:
@jayfoad But the implicit write to VCC doesn't "depend on exec", right? Should we adjust the comment?

Reply:
> Probably should directly check the first operand is the only explicit def rather than assume it's implied by the encoding

I have adjusted the condition. The downside is that it's less clear to me now if returning false is correct for all instructions covered by this condition.

Reply:
Before the [commit](https://github.com/llvm/llvm-project/commit/a1fb307b4b8d5ffd8f66a477783e362e54a3c03c) by @perlfu  which introduced the logic that tries to "ignore comparisons which are only used masked with exec", this function was returning true for all comparisons. Hence returning true in our case should be acceptable, too.

Reply:
Conservatively returning `true` for this test is always safe.
The original point was to allow hoisting of more instructions, returning true basically means "don't hoist".

Reply:
It's depends on exec in a boolean way, not as a data value way. i.e. single lane bit value vs. lane mask